zvariant/src/file_path.rs

```diff
@@ -0,0 +1,106 @@
+use std::{ffi::{OsStr, OsString}, path::{Path, PathBuf}};
+
+use serde::{de::{self, Visitor}, Deserialize, Deserializer, Serialize, Serializer};
+use crate::de::Deserializer as ZbusDeserializer;
+
+use crate::Type;
+
+/// A file name represented as a nul-terminated byte array.
+#[derive(Type, Debug, Default, PartialEq)]
+#[zvariant(signature = "ay")]
+pub struct FilePath(OsString);
+
+impl From<&Path> for FilePath {
+    fn from(value: &Path) -> Self {
+        Self(value.as_os_str().to_os_string())
+    }
+}
+
+impl From<PathBuf> for FilePath {
+    fn from(value: PathBuf) -> Self {
+        Self(value.as_os_str().to_os_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for FilePath {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>
+    {
+        struct FilePathVisitor;
+        impl<'de> Visitor<'de> for FilePathVisitor {
+            type Value = FilePath;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                formatter.write_str("unable to deserialize FilePath")
+            }
+
+            fn visit_bytes<E>(self, v: &[u8]) -> std::result::Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                unsafe {
+                    Ok(FilePath(
+                            OsStr::from_encoded_bytes_unchecked(v).to_os_string()
+                    ))
```

Comment:
Hmm.. Are we safe here? The docs say that we're violating the safety rules here:

> As the encoding is unspecified, callers must pass in bytes that originated as a mixture of validated UTF-8 and bytes from [OsStr::as_encoded_bytes](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.as_encoded_bytes) from within the same Rust version built for the same target platform. For example, reconstructing an OsStr from bytes sent over the network or stored in a file will likely violate these safety rules.

In any case, we'll want a `Safety: ` commend above that proves why this is safe.

Reply:
- It was an assumption rather than an assertion i.e. not **necessarily,** but **safe to assume** as a path can contain other non-UTF8 symbols, so we have no reason to say that it only contains UTF8 symbols, thus, the assumption.
- yes, you were right, the comment has nothing to do with the `unsafe` call, by bad ig ;p
- Also I've moved out from the unsafe call to a [more safe approach](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_vec_with_nul)

Reply:
> * It was an assumption rather than an assertion i.e. not **necessarily,** but **safe to assume** as a path can contain other non-UTF8 symbols, so we have no reason to say that it only contains UTF8 symbols, thus, the assumption.

Right but assuming UTF-8 where bytes could have non-UTF8 would be problematic. The other way around is not a problem and hence need no documentation and only makes it confusing.

> * yes, you were right, the comment has nothing to do with the `unsafe` call, by bad ig ;p

No worries. That's why we've reviews.

> * Also I've moved out from the unsafe call to a [more safe approach](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_vec_with_nul)

Cool but seems all it really gives us is a null-byte check (which we can do safely ourselves too). I also realised that we're forcing copying/allocation where we probably don't need to. I think you should be implementing/using `Visitor::visit_borrowed_bytes` instead combined with https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.from_encoded_bytes_unchecked .

I'm still a bit unsure about safety of it all here. For example, what happens if the invariant of these methods don't hold. Maybe we can keep this type unix-only, since at least for unix platforms, you can safely deserialize `&OsStr` from bytes: https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html

Reply:
Hi! I've made some changes that you can find [here](https://gist.github.com/yassinebenarbia/aa0c536eaa364d3985e35edb6153736c), I've tried to minimize copying/allocating memory as much as I can tho, but I'm not quite sure that I didn't miss something, so please let me know if you spotted anything :p, also, if you find those changes satisfying, I'll be adding them to the PR

Reply:
Cool. Can you please just push it to this branch so it's easier to review specific parts of the code?